src/services/osu-service.js

```diff
@@ -238,18 +238,30 @@ export default class OSUService {
     parser.feed(mapString);
     const { map } = parser;
     // eslint-disable-next-line new-cap
-    const stars = new OSU.diff().calc({
-      map,
-      mods: +enabledMods
-    });
-    const pp = OSU.ppv2({
-      stars,
-      combo: +maxcombo,
-      nmiss: +countmiss,
-      n50: +count50,
-      n100: +count100,
-      n300: +count300
-    });
+
+    try { //FIXED : NotImplementedError when calculate diff or pp for unsupported game modes
+        const stars = new OSU.diff().calc({
+            map,
+            mods: +enabledMods
+        });
+        const pp = OSU.ppv2({
+            stars,
+            combo: +maxcombo,
+            nmiss: +countmiss,
+            n50: +count50,
+            n100: +count100,
+            n300: +count300
+        });
+    } catch (e) {
+        // Return when catch a NotImplementedError
+        if (e.name === "NotImplementedError")
+            return {
+                acc: 0,
+                pp: "N/A",
+                map
+            };
+      }
```

Comment:
getPP 的调用方有个如果为 string 类型返回值就直接输出的逻辑 👇
https://github.com/zcWSR/japari-admin/blob/ab8847b06b226cda1f619b28a4d967368dda2cda/src/services/osu-service.js#L266-L272

所以这里其实可以改成返回 string，并且如果判断不出错误可以考虑继续向上抛，顶层 koa 有个 middleware 会拦截住异常

Reply:
修改建议
```suggestion
        if (e.name === "NotImplementedError") {
            return 'mode 不支持';
        }
        new Error('未知错误');
      }
```

Reply:
这里意图是在遇到不支持的模式时，也能够正常输出大部分数据。所以返回一个人为指定的对象。
```suggestion
        if (e.name === "NotImplementedError")
            return {
                acc: 0,
                pp: "N/A",
                map
            };
            throw e;
      }
```